src/driver-holtek/holtek8-shared.c

```diff
@@ -0,0 +1,1291 @@
+
+unsigned int
+holtek8_raw_to_report_rate(uint8_t raw)
+{
+	const struct holtek8_report_rate_mapping *mapping = NULL;
+	ARRAY_FOR_EACH(holtek8_report_rate_map, mapping)
+		if (mapping->raw == raw)
+			return mapping->report_rate;
+	return 0;
+}
+
+uint16_t
+holtek8_dpi_to_raw(struct ratbag_device *device, unsigned int dpi)
+{
+	struct holtek8_data *drv_data = device->drv_data;
+	const struct holtek8_sensor_config *sensor_cfg = drv_data->sensor_cfg;
+
+	if (dpi < sensor_cfg->dpi_min)
+		dpi = sensor_cfg->dpi_min;
+
+	if (dpi > sensor_cfg->dpi_max)
+		dpi = sensor_cfg->dpi_max;
+
+	uint16_t raw = dpi / sensor_cfg->dpi_step;
+
+	if (sensor_cfg->zero_indexed)
+		raw -= 1;
+
+	return raw;
+}
+
+unsigned int
+holtek8_raw_to_dpi(struct ratbag_device *device, uint16_t raw)
+{
+	struct holtek8_data *drv_data = device->drv_data;
+	const struct holtek8_sensor_config *sensor_cfg = drv_data->sensor_cfg;
+
+	if (sensor_cfg->zero_indexed)
+		raw += 1;
+
+	return raw * sensor_cfg->dpi_step;
+}
+
+static bool
+holtek8_button_data_is_equal(const struct holtek8_button_data *lhs, const struct holtek8_button_data *rhs)
+{
+	if (lhs->type != rhs->type)
+		return false;
+
+	for (unsigned int i = 0; i < sizeof(lhs->data); ++i) {
+		if (lhs->data[i] != rhs->data[i]) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static bool
+holtek8_button_map_action_to_raw(const struct ratbag_button_action *action, struct holtek8_button_data *data)
+{
+	const struct holtek8_button_mapping *mapping = NULL;
+
+	ARRAY_FOR_EACH(holtek8_button_map, mapping) {
+		if (!ratbag_button_action_match(&mapping->action, action)) {
+			continue;
+		}
+
+		*data = mapping->data;
+		return true;
+	}
+
+	return false;
+}
+
+static const struct ratbag_button_action *
+holtek8_button_map_raw_to_action(const struct holtek8_button_data *data)
+{
+	const struct holtek8_button_mapping *mapping = NULL;
+
+	ARRAY_FOR_EACH(holtek8_button_map, mapping) {
+		if (!holtek8_button_data_is_equal(data, &mapping->data))
+			continue;
+
+		return &mapping->action;
+	}
+
+	return NULL;
+}
+
+struct holtek8_modifier_mapping {
+	enum holtek8_modifiers modifiers;
+	unsigned int key;
+};
+
+static const struct holtek8_modifier_mapping holtek8_modifier_map[] = {
+	{ HOLTEK8_MODIFIER_LEFTCTRL, KEY_LEFTCTRL },
+	{ HOLTEK8_MODIFIER_LEFTSHIFT, KEY_LEFTSHIFT },
+	{ HOLTEK8_MODIFIER_LEFTALT, KEY_LEFTALT },
+	{ HOLTEK8_MODIFIER_LEFTMETA, KEY_LEFTMETA },
+	{ HOLTEK8_MODIFIER_RIGHTCTRL, KEY_RIGHTCTRL },
+	{ HOLTEK8_MODIFIER_RIGHTSHIFT, KEY_RIGHTSHIFT },
+	{ HOLTEK8_MODIFIER_RIGHTALT, KEY_RIGHTALT },
+	{ HOLTEK8_MODIFIER_RIGHTMETA, KEY_RIGHTMETA },
+};
+
+/*
+ * Reads macro events from the device memory.
+ *
+ * Macros in memory are divided into pages. If the device supports jumping between
+ * pages, tries to follow jumps to read multipaged macros.
+ *
+ * @return 0 on success or a negative errno.
+ */
+static int
+holtek8_read_macro_data(struct ratbag_device *device, struct holtek8_macro_event *macro_events, uint8_t macro_idx)
```

Comment:
Please, pass along 'macro_events' size instead of expecting it to be equal to 'HOLTEK8_MAX_MACRO_EVENTS'.